### PR TITLE
chore: set up auto for versioning/release management

### PR DESCRIPTION
**Pull request by hydrosquall**
## Motivation

- Dry-run of an automated release process before using it elsewhere in the vega ecosystem
- Maintainers spend more time on Dataviz and less time on repository maintenance (versioning, making changelogs, publishing, testing changes by having to build locally, etc). Based on peer work in https://github.com/nteract/data-explorer/pull/63

## Changes

- Adds + configures `auto` . This way, you get hinting/docs if editing locally.
- Renames the job, otherwise it doesn't run by default since I wasn't a maintainer when I first opened the PR
- ~Does not remove semantic-release or related config until we confirm this is working.~ (It's [ok](https://github.com/vega/ts-json-schema-generator/pull/856#issuecomment-881114170))
- Ran `auto create-labels` locally to seed repository with library labels
Image 2021-07-15 at 9 10 27 PM [link](https://a.cl.ly/9Zudnrqx)

## Testing

- Succeeded in : https://github.com/vega/ts-json-schema-generator/runs/3082312581?check_suite_focus=true
- Workflow will publish a canary release if run by a maintainer,  the PR has a release, and [`auto create-labels`](https://intuit.github.io/auto/docs/generated/create-labels) has been run once locally. 
  - If all goes well, there will be an edit on this PR description with a link to the canary deploy.
- This is a redo of #855 with correct repo permissions.

## Notes on comparison vs semantic-release

- `auto` supports opening per-PR canary deploys rather than depending on merges to a release branch
- Philosophically, auto is encouraging releasing all the time (even non-semantic releases)

<details>
  <summary>📦 Published PR as canary version: <code>0.94.2--canary.856.e019c4a.0</code></summary>
  <br />
  
  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install ts-json-schema-generator@0.94.2--canary.856.e019c4a.0
  # or 
  yarn add ts-json-schema-generator@0.94.2--canary.856.e019c4a.0
  ```
</details>

### CODE REVIEW

File: .autorc.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "plugins": [
+    "npm",
+    "conventional-commits",
+    "first-time-contributor",
+    "released"
+  ],
+  "onlyPublishWithReleaseLabel": true
```
Comment (hydrosquall) [inline .autorc.json:8]: @domoritz , if we remove this boolean, the library will auto-publish on every merge to `master`. I suggest we don't enable that until you're first happy with how the auto-version incrementing logic works (it depends on committers adhering well to git commit message guidelines). 

There are other options for setting version (setting a major/minor/patch github label), or asking users to check off the appropriate checkbox in their PR description when opening the PR, but this is probably the easiest place to start.

Comment (domoritz) [reply .autorc.json:8]: I was wondering about all the labels that this process added. I am pretty strict with enforcing correct pull request descriptions withs semantic commits. Can we use them to automatically set the right version? Or is the idea of the labels that they get added automatically?

**Comment by domoritz** [comment]
Btw, I'm happy to remove the semantic-release dependency in this pull request as well.

**Comment by domoritz** [comment]
How do canary releases work if the pull request comes from external (as most prs do)? I suppose they won't happen in that case.

File: .autorc.json (hunk 1 above)
Comment (hydrosquall) [reply .autorc.json:8]: > Can we use them to automatically set the right version?

I wasn't sure about the semantic commit situation in this repo because I didn't see commit-hooks. However, I think I saw some linting config in a different vega repo. Whether having good commit messages is checked by eye or tool, https://github.com/nteract/data-explorer/pull/66, having manual labeling as an option is just a fallback for inaccurate messages. 

The automatic labeling is based on `conventional-commits`, which most likely has the same behavior as the semantic-release plugin.

https://intuit.github.io/auto/docs/generated/conventional-commits

If you wanted to do the manual checkbox thing, it could be achieved with this plugin:

https://intuit.github.io/auto/docs/generated/pr-body-labels

**Comment by hydrosquall** [comment]
> How do canary releases work if the pull request comes from external (as most prs do)? I suppose they won't happen in that case.

I think they won't run if the external person is a 1st time contributor, per https://github.blog/changelog/2021-04-22-github-actions-maintainers-must-approve-first-time-contributor-workflow-runs/, but after an initial approval, I think newcomers would have the ability to publish canaries.

File: package.json [hunk 2]
```diff
@@ -88,23 +89,7 @@
...
-      "@semantic-release/release-notes-generator",
-      "@semantic-release/github",
-      "@semantic-release/npm",
-      [
-        "@semantic-release/git",
-        {
-          "assets": [
-            "package.json"
-          ],
-          "message": "chore(release): ${nextRelease.version}"
```
Comment (hydrosquall) [inline package.json:105]: Version bumps by default look like this : https://github.com/nteract/data-explorer/commit/252d43adf0551ec97eed4c8418f51d3ece247632

File: .autorc.json (hunk 1 above)
Comment (hydrosquall) [reply .autorc.json:8]: Last link: if any projects have more specific labeling needs than the defaults (or different mappings in terms of what we consider major/minor), they can be customized by passing in a custom array in the config:

https://intuit.github.io/auto/docs/configuration/autorc#labels

Comment (domoritz) [reply .autorc.json:8]: Here is how we ensure commit messages:

Screen Shot 2021-07-15 at 21 34 53

I don't want additional labels if they are not necessary.

**Comment by domoritz** [comment]
> but after an initial approval, I think newcomers would have the ability to publish canaries.

Secrets are not exposed to forks so not npm releases for them. 

Screen Shot 2021-07-15 at 21 37 23

File: .autorc.json (hunk 1 above)
Comment (hydrosquall) [reply .autorc.json:8]: Oh neat I missed the `semantic pull request` integration! That's even more reliable than a commit hook since it doesn't depend on someone setting up local config properly. I might set that up for another project, thanks for the tip.

Regarding labels: I *think* we can delete the unused tags then. The other ones are optional and by default would help organize the auto-generated changelog, but are not required. I just deleted most of the added tags, with the exception of `release` and `released` since those are important for keeping track of what has been deployed. 

- https://intuit.github.io/auto/docs/welcome/getting-started#3-create-labels

**Comment by hydrosquall** [comment]
> > but after an initial approval, I think newcomers would have the ability to publish canaries.
> 
> Secrets are not exposed to forks so not npm releases for them.
> 
> Screen Shot 2021-07-15 at 21 37 23

Ah, good to know. I guess the primary initial beneficiaries of canary builds would be project maintainers. I think this limitation makes sense as a way to reduce the risk of abuse (e.g. people spamming your NPM account with bad builds), but we can still help give people a fast feedback loop if we decide to treat `master` as a pre-release branch (so it publishes everything using the command [next](https://intuit.github.io/auto/docs/generated/next)), and make a new branch like `stable` or similar for "official" releases. 

Let me know if you'd like to pursue that option (adding a new branch for stable, and treating `master` as a source of pre-releases) in this PR, or save that for a future one.

**Comment by domoritz** [comment]
Nah, I want master to be the branch everything goes into and make releases from time to time. 

So, if we accept this pull request, what happens to prs from forks? Would the ci fail since the NPM token and GitHub token don't exist? 

How do I make a release now? I used to call a particular GitHub action to kick off a release. How do I make a release after all prs are merged? Or does a release always happen with a pull request?

**Comment by hydrosquall** [comment]
> So, if we accept this pull request, what happens to prs from forks? Would the ci fail since the NPM token and GitHub token don't exist?

Good point. We have at least 2 options

1. Check for presence of the environment variable, quit job early if not present
2. Quit if the repository context isn't `vega//ts-json-schema-generator` https://github.community/t/stop-github-actions-running-on-a-fork/17965/2 . 

Let me know which one you prefer.

> How do I make a release now? I used to call a particular GitHub action to kick off a release. How do I make a release after all prs are merged? Or does a release always happen with a pull request?

As it's currently set up, the workflow will release every merged PR, since the action is triggered on `push`. If you'd prefer to manually make releases, we'd assign a different trigger to the github action.

You can check to confirm you like what the command would do on the main branch by running `yarn run auto latest --dry-run` (https://intuit.github.io/auto/docs/generated/shipit). Currently since we're running on a feature branch, we're seeing the results of `yarn run auto canary`.

**Comment by domoritz** [comment]
Let me rephrase. I am okay with canary releases with every pull request but I don't want releases where I have to promise backwards compatibility with every merged pull request. I want to manually decide when we make a release. Bonus points if I don't have to do anything locally. 

It would be great to document the process a little bit in the github contributing document.

**Comment by hydrosquall** [comment]
> I am okay with canary releases with every pull request but I don't want releases where I have to promise backwards compatibility with every merged pull request. I want to manually decide when we make a release. Bonus points if I don't have to do anything locally.

What you describe makes sense. I propose that we follow the 2 branch scheme (see https://intuit.github.io/auto/docs/generated/shipit#next-branch-default for a picture + 4 sentence description )

 I did a proof of concept of what I suggest using a small test project (https://github.com/hydrosquall/storybook-addon-datadog-rum/pull/9#issue-691800646) 

Here's my proposal. If it sounds good to you to try, I'll write up a doc and add it to this PR (but feel free to ask for further questions or clarification).

### Reasoning

*Problem*: It's valuable to automatically release the result of every merged PR approved so maintainers don't have to manually cut releases just because a 1st time contributor opened a change. However, preserving stability within semantic ranges on the `latest` tag is important, because not every PR is going to warrant bumping the version on the "latest" channel. 

*Solution*: 

- Release everything on the `next` tag for those who want to live on the cutting-edge
- Manually open a PR from `next` to `main` when you're ready to open a stable release

## Method

To get the best of all worlds, I suggest we 

  1. Create a `next` branch as our pre-release branch. All PRs should be made here.
  2. Create a `stable` branch as the branch which is used for stable deploys.
  3. Make `next` the repository default branch so that all new PRs are opened into it
  4. Whenever you want to manually cut a release (nothing local needs to be done), open a PR from `next` into `stable` with whatever title you want. The version bump will be calculated by conventional-commits (see [docs](https://intuit.github.io/auto/docs/generated/conventional-commits) ). 

When all open PRs made against it are closed, I also suggest we remove the `master` branch so people don't get confused (but it's OK to leave it around).

**Comment by hydrosquall** [comment]
Per the slack thread, I initially added the `magic-zero` plugin, but it doesn't work. Added a comment with some debugging output for the library authors: https://github.com/intuit/auto/issues/1946#issuecomment-881969396

**Comment by domoritz** [comment]
Thank you for explaining the process I think I like the two branches approach.

For apps like the editor, we could also have two branches and auto release the stable branch (maybe call it deployed). 

Can you automatically disable canary releases for dependably updates? I think that's getting too noisy otherwise.

Paging @kanitw and @arvind to see whether they agree and want us to adopt this approach in Vega-Lite, embed, themes, tooltips etc as well.

**Comment by hydrosquall** [comment]
> Can you automatically disable canary releases for dependably updates? I think that's getting too noisy otherwise.

I think so, would you want to stop just canary releases (the release that shows up in the comment PR)  or also pre-releases (what gets triggered after merging into `next`)? This would be handled at the github actions level rather than something internal to shipit (if we need more specialized logic, we could write a shipit plugin.

https://github.community/t/how-to-stop-github-actions-workflow-to-trigger-when-the-pull-request-is-from-dependabot-preview/116486/2

**Comment by hydrosquall** [comment]
Updated base branch from `master` to `next`, after this is merged, we should edit the repo default branch from `main` to `next` so that new PRs will be made relative to it instead.

File: auto.config.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        plugins: [
+            "npm",
+            "conventional-commits",
+            "first-time-contributor",
+            "released"
+        ],
+        // Follow the 2 branch deployment scheme
+        // https://intuit.github.io/auto/docs/generated/shipit#next-branch-default
+        baseBranch: "stable",        // latest "official" release
+        prereleaseBranches: ["main"] // latest changes (subject to breaking). next, alpha, beta, and other multi-feature test branches can all be added here.
```
Comment (hydrosquall) [inline auto.config.ts:16]: Haven't edited this config yet, but plan to make switch this to `next` after confirming it makes sense w/ @domoritz to rotate this to the default branch

Comment (domoritz) [reply auto.config.ts:16]: Rotating the default branch makes sense to me.

File: .github/workflows/publish-auto.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+on: [push]
+  branches-ignore:
+    # No canary deploys for branches opened by dependabot
+    - 'dependabot/**'
+
+jobs:
+  publish:
+    runs-on: ubuntu-latest
+    if: "!contains(github.event.head_commit.message, 'ci skip') && !contains(github.event.head_commit.message, 'skip ci')"
```
Comment (domoritz) [inline .github/workflows/publish-auto.yml:11]: Do we use this right now? If not, let' remove it.

File: auto.config.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { AutoRc } from "auto";
+
+
+/** Auto configuration */
+export default function rc(): AutoRc {
+    return {
+        plugins: [
+            "npm",
+            "conventional-commits",
+            "first-time-contributor",
```
Comment (domoritz) [inline auto.config.ts:10]: What does this do?

**Review by domoritz — APPROVED** [review]
Looks good to me. Let's do it (please squash the commits in here when merging). 

The next steps are updating the following repos with the same proess

* vega-lite-dev-config
* vega-embed
* vega-tooltip
* vega-themes
* vega-datasets
* vega-lite

And then Vega if @jheer wants it.

File: .github/workflows/publish-auto.yml (hunk 4 above)
Comment (hydrosquall) [reply .github/workflows/publish-auto.yml:11]: This is necessary because auto create commits which contain `ci skip)` when it updates the changelog and bumps the version in the `stable` branch. This guard is present to prevent auto from triggering an infinite loop (creating a commit, rerunning auto, etc)

Image 2021-07-20 at 9 14 22 PM

https://intuit.github.io/auto/docs/build-platforms/github-actions

File: auto.config.ts (hunk 5 above)
Comment (hydrosquall) [reply auto.config.ts:10]: When someone contributes to a repository for the first time, it tags and thanks them in the first [Changelog heading](https://github.com/nteract/data-explorer/blob/main/CHANGELOG.md#v8210-tue-jun-29-2021) and [Release Notes](https://github.com/nteract/data-explorer/releases/tag/v8.2.10) , and doesn't do anything if this release doesn't have any 1st time contributors. 

Docs: https://intuit.github.io/auto/docs/generated/first-time-contributor

We can remove this if it wouldn't be relevant for the Vega projects, but I figured it couldn't hurt + might help with encouraging people to become ongoing contributors.

File: auto.config.ts (hunk 3 above)
Comment (hydrosquall) [reply auto.config.ts]: Updated to `next`: https://github.com/vega/ts-json-schema-generator/pull/856/commits/492820bcf73b271a1a4c233031be8a41b84e6717

I don't have admin permissions on this repo, so I'll need your help to switch the default branch to next. I just synced next to the HEAD of `master` (`8439cfbb45918c11ac5df65e9a516a07763478a9`) so we should be ready to go.

**Comment by hydrosquall** [comment]
> Looks good to me. Let's do it (please squash the commits in here when merging).

Will do! By the way, within the repository settings, we can actually require squash commits for repositories if we want to prevent accidental misclicks: https://docs.github.com/en/github/administering-a-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

> The next steps are updating the following repos with the same process

Sounds good, I made a card to track the migrations. I'll move these over gradually, I remember you pointed out that some of these repos may have more elaborate build processes than others

https://github.com/orgs/vega/projects/8?add_cards_query=is%3Aopen

**Comment by domoritz** [comment]
Vega-Lite will be the trickiest. Have a look at the current scripts and let me know about any questions you have.

**Comment by domoritz** [comment]
> By the way, within the repository settings, we can actually require squash commits for repositories if we want to prevent accidental misclicks

I found that dependabot updates often have fewer conflicts when I rebase. Maybe that's not true anymore but that's why I left support for rebase on.
